cpp/src/arrow/python/datetime.cc

```diff
@@ -265,21 +255,25 @@ static inline Status PyDateTime_from_int(int64_t val, const TimeUnit::type unit,
   return Status::OK();
 }
 
-static inline int64_t PyDate_to_days(PyDateTime_Date* pydate) {
-  return get_days_from_date(PyDateTime_GET_YEAR(pydate), PyDateTime_GET_MONTH(pydate),
-                            PyDateTime_GET_DAY(pydate));
+using TimePoint =
+    std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>;
+
+Status PyDateTime_from_TimePoint(TimePoint val, PyObject** out) {
+  auto nanos = val.time_since_epoch();
+  auto micros = std::chrono::duration_cast<std::chrono::microseconds>(nanos);
+  if (micros < nanos) {
+    micros += std::chrono::microseconds(1);  // ceil
```

Comment:
Hmm, but why?

Reply:
Tried to approximate the python behaviour, I can fall back to truncation.

Reply:
Actually... Python doesn't return a `datetime` in `pathlib`, just a float of the timestamp in seconds. So we should just do the same IMO.

Reply:
I've removed the ceiling, but we still return with a datetime. I'd prefer datetime, but I can change it.

Reply:
Following Python sounds preferable to me. Especially as `datetime` loses precision.